### PR TITLE
Sais unsupervised

### PR DESCRIPTION
**Pull request by sainag7**

### CODE REVIEW

**Review by VitorInserra — COMMENTED** [review]
Remove LSTM-based encoders.

Instead, add LSTM and BiLSTM to supervised architectures, with a modular loss function parameter. Use softmax cross-entropy until Variational loss is complete.

File: src/eegproc/unsupervised/unsupervised.py [hunk 1]
```diff
@@ -385,3 +403,240 @@ def encoder_gcn(
     seq_emb = layers.Dense(emb_dim, activation=None, name="seq_emb")(x)
 
     return Model(inputs=x_in, outputs=seq_emb, name="encoder_gcn")
+
+
+def encoder_lstm(
```
Comment (VitorInserra) [inline src/eegproc/unsupervised/unsupervised.py]: No need for LSTM encoder architectures. RNNs haven't shown much success in dimensionality reduction for EEG data.

File: src/eegproc/unsupervised/unsupervised.py [hunk 2]
```diff
@@ -385,3 +403,240 @@ def encoder_gcn(
...
+        if include_softmax:
+            output = layers.Softmax(name="class_probabilities")(x)
+        else:
+            output = x  # raw logits, no softmax
+        model_name = "encoder_lstm_classifier"
+
+    return Model(inputs=x_in, outputs=output, name=model_name)
+
+
+def encoder_bilstm(
```
Comment (VitorInserra) [inline src/eegproc/unsupervised/unsupervised.py]: Same here!

**Review by Pranav1006 — COMMENTED** [review]
Everything looks well documented, designed, and implemented!

**Comment by Pranav1006** [comment]
Also, I find the copilot suggested changes very beneficial. I highly suggest you use them (copilot can fix and commit for you if you just ask it) since it spots things that a lot of us don't notice.

File: src/eegproc/unsupervised/cross_validation.py [hunk 3]
```diff
@@ -0,0 +1,1024 @@
...
+
+Four strategies are implemented:
+
+    LOSO CV         Leave One Subject Out
+                    Hold out one subject at a time as the test set.
+                    This is the gold standard for testing cross-subject
+                    generalization. Number of folds = number of subjects.
+
+    LOO CV          Leave One Out (one sample/window at a time)
+                    Hold out a single sample window at a time. Very
```
Comment (VitorInserra) [inline src/eegproc/unsupervised/cross_validation.py:16]: This should not be a single sample window, this should be a single session or trial and should be identified by a selected column parameter like id or session_id (optional to dev).

File: src/eegproc/unsupervised/cross_validation.py [hunk 4]
```diff
@@ -0,0 +1,1024 @@
...
+                    fine-grained, produces many folds, and tests
+                    within-dataset generalization. Best used on smaller
+                    datasets where LOSO would leave too little training data.
+
+    LKOCV           Leave K Subjects Out
```
Comment (VitorInserra) [inline src/eegproc/unsupervised/cross_validation.py:21]: LKO is Leave-K-Out, was defined in our previous work as a session/trial level holdout. What you wrote here is actually LNSO.

File: src/eegproc/unsupervised/cross_validation.py [hunk 5]
```diff
@@ -0,0 +1,1024 @@
...
+        feature_array=X,
+        label_array=y,
+        subject_id_array=subject_ids,
+        n_epochs=30,
+        batch_size=16,
+    )
+
+    print(results["mean_scores"])   # average accuracy and loss across folds
+"""
+
```
Comment (VitorInserra) [inline src/eegproc/unsupervised/cross_validation.py:81]: Major request:

Change the current LKO CV to LNSO CV (simply change the name).

Change LOO CV to LKOCV (EEG sessions don't make sense with LOOCV usually). LKO CV will be our trial level hold hout. 

Add Nested-LKO CV.

Make LOSO CV simply call LNSO CV with N=1

File: src/eegproc/unsupervised/cross_validation.py [hunk 6]
```diff
@@ -0,0 +1,1024 @@
...
+    batch_size: int = 32,
+    verbose: int = 0,
+    extra_fit_kwargs: dict | None = None,
+) -> dict:
+    """Leave One Subject Out Cross Validation (LOSO CV).
+
+    In each fold one subject is completely held out as the test set and the
+    model is retrained from scratch on all remaining subjects. Because the
+    test subject's data is *never* seen during training, this strategy
+    measures true cross-subject generalization — the most rigorous EEG
```
Comment (VitorInserra) [inline src/eegproc/unsupervised/cross_validation.py]: This is an overclaim. Cross-subject generalization is also more robust in LNSO or N LNSO,

File: src/eegproc/unsupervised/cross_validation.py [hunk 7]
```diff
@@ -0,0 +1,1024 @@
...
+# ---------------------------------------------------------------------------
+# LOSO CV — Leave One Subject Out Cross Validation
+# ---------------------------------------------------------------------------
+
+def loso_cv(
+    model_builder_function: Callable[[], tf.keras.Model],
+    feature_array: np.ndarray,
+    label_array: np.ndarray,
+    subject_id_array: np.ndarray,
+    n_epochs: int = 50,
```
Comment (VitorInserra) [inline src/eegproc/unsupervised/cross_validation.py]: This won't work with models that don't use epochs or batch_size. This should either require an args dictionary of hyperparams or have the hyperparams defined before passing. Also, think about how this will be compatible with optuna.

Pass a dict of lists of hyperparameters and treat that as a grid search for now. The model should then go through them. We need to define standard arguments for VAE CNN and GCN paired with LSTM and BiLSTM. Maybe we will need to break up the CV functions.

File: src/eegproc/unsupervised/cross_validation.py [hunk 8]
```diff
@@ -0,0 +1,1024 @@
...
+        # Slice the data into training and test portions.
+        X_train = feature_array[train_indices]
+        y_train = label_array[train_indices]
+        X_test  = feature_array[test_indices]
+        y_test  = label_array[test_indices]
+
+        # Build a fresh model for this fold (no weight leakage from prior folds).
+        model = model_builder_function()
+
+        model.fit(
```
Comment (VitorInserra) [inline src/eegproc/unsupervised/cross_validation.py]: Unclear what is constituting the validation set within optimization. We need to partition the dataset in three: training/validation/test. Validation serves for validation loss. If validation size (n_users) is not defined then we need to fit to within the training set. This has to be handled here to reduce margin for error in the training functions.

**Review by VitorInserra — CHANGES_REQUESTED** [review]
Good work. We will have to iteratively enhance this, it is an essential but difficult portion of the project and framework and a part that many BCI researchers get wrong. I requested a few changes, let's start just by rewriting `LOSO CV`, then when we get that down we can move on to the next ones.
